# Benchmark
[Text Embedding](https://python.langchain.com/docs/integrations/text_embedding/)

In [1]:
import tracemalloc
import subprocess
import os

folder_name = "Local"

if not os.path.exists(folder_name):
    os.makedirs(folder_name)

    subprocess.run(["git", "clone", "https://github.com/github/docs.git", folder_name])

In [2]:
import markdown_splitter as mdS
from langchain_community.vectorstores import FAISS

docs = mdS.create_db(path=folder_name, glob="**/*.md",split=None)

# data/reusables/open-source/open-source-guide-general.md
query = "Where can I find guidance on creating and nurturing an open source project"

100%|█████████▉| 5926/5931 [00:01<00:00, 4431.13it/s]


In [3]:
# print(len(docs))
# docs = docs[0:2000]

# BGE  & FAISS

In [4]:
'''
tracemalloc.start()
print(tracemalloc.get_traced_memory())

# Embeddings
from langchain_community.embeddings import HuggingFaceBgeEmbeddings
embeddings = HuggingFaceBgeEmbeddings(model_name="BAAI/bge-small-en", model_kwargs={"device": "cpu"}, encode_kwargs={"normalize_embeddings": True})

# Store
docsearch = await FAISS.afrom_documents(docs, embeddings)

print(tracemalloc.get_traced_memory())
tracemalloc.stop()
'''

'\ntracemalloc.start()\nprint(tracemalloc.get_traced_memory())\n\n# Embeddings\nfrom langchain_community.embeddings import HuggingFaceBgeEmbeddings\nembeddings = HuggingFaceBgeEmbeddings(model_name="BAAI/bge-small-en", model_kwargs={"device": "cpu"}, encode_kwargs={"normalize_embeddings": True})\n\n# Store\ndocsearch = await FAISS.afrom_documents(docs, embeddings)\n\nprint(tracemalloc.get_traced_memory())\ntracemalloc.stop()\n'

In [5]:
'''
%%time
for i in range(10):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)
'''

'\n%%time\nfor i in range(10):\n    docsearch.similarity_search(query)\n\ndocsearch.similarity_search(query)\n'

# FastEmbed & FAISS

In [6]:
'''
tracemalloc.start()
print(tracemalloc.get_traced_memory())

# Embeddings
from langchain_community.embeddings.fastembed import FastEmbedEmbeddings
embeddings = FastEmbedEmbeddings()

# Store
docsearch = await FAISS.afrom_documents(docs, embeddings)

print(tracemalloc.get_traced_memory())
tracemalloc.stop()
'''

'\ntracemalloc.start()\nprint(tracemalloc.get_traced_memory())\n\n# Embeddings\nfrom langchain_community.embeddings.fastembed import FastEmbedEmbeddings\nembeddings = FastEmbedEmbeddings()\n\n# Store\ndocsearch = await FAISS.afrom_documents(docs, embeddings)\n\nprint(tracemalloc.get_traced_memory())\ntracemalloc.stop()\n'

In [7]:
'''
%%time
for i in range(10):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)
'''

'\n%%time\nfor i in range(10):\n    docsearch.similarity_search(query)\n\ndocsearch.similarity_search(query)\n'

# GPT4All & FAISS

In [8]:
tracemalloc.start()
print(tracemalloc.get_traced_memory())

# Embeddings
from langchain_community.embeddings import GPT4AllEmbeddings
embeddings = GPT4AllEmbeddings()

# Store
docsearch = await FAISS.afrom_documents(docs, embeddings)

print(tracemalloc.get_traced_memory())
tracemalloc.stop()

(928, 11633)
bert_load_from_file: gguf version     = 2
bert_load_from_file: gguf alignment   = 32
bert_load_from_file: gguf data offset = 695552
bert_load_from_file: model name           = BERT
bert_load_from_file: model architecture   = bert
bert_load_from_file: model file type      = 1
bert_load_from_file: bert tokenizer vocab = 30522
(247281963, 282854199)


In [9]:
%%time
for i in range(10):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 2.92 s, sys: 40.4 ms, total: 2.96 s
Wall time: 838 ms


[Document(page_content="For more information on open source, specifically how to create and grow an open source project, we've created [Open Source Guides](https://opensource.guide/) that will help you foster a healthy open source community."),
 Document(page_content="For more information on open source, specifically how to create and grow an open source project, we've created [Open Source Guides](https://opensource.guide/) that will help you foster a healthy open source community by recommending best practices for creating and maintaining repositories for your open source project."),
 Document(page_content='There are a variety of ways that you can contribute to open source projects.', metadata={'Header 2': 'Validating an issue or pull request'}),
 Document(page_content='- "[Your Code of Conduct](https://opensource.guide/code-of-conduct/)" from the Open Source Guides\n- "[Building Welcoming Communities](https://opensource.guide/building-community/)" from the Open Source Guides\n- "[Lea

# Hugging Face & FAISS

In [3]:
tracemalloc.start()
print(tracemalloc.get_traced_memory())

# Embeddings
from langchain_community.embeddings import HuggingFaceEmbeddings
embeddings = HuggingFaceEmbeddings()

# Store
docsearch = await FAISS.afrom_documents(docs, embeddings)

print(tracemalloc.get_traced_memory())
tracemalloc.stop()

(928, 11633)
(611073929, 667287104)


In [4]:
%%time
for i in range(10):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 2.41 s, sys: 921 ms, total: 3.33 s
Wall time: 921 ms


[Document(page_content="For more information on open source, specifically how to create and grow an open source project, we've created [Open Source Guides](https://opensource.guide/) that will help you foster a healthy open source community."),
 Document(page_content="For more information on open source, specifically how to create and grow an open source project, we've created [Open Source Guides](https://opensource.guide/) that will help you foster a healthy open source community by recommending best practices for creating and maintaining repositories for your open source project."),
 Document(page_content='Contributing to open source projects on {% data variables.product.prodname_dotcom %} can be a rewarding way to learn, teach, and build experience in just about any skill you can imagine. For more information, see "[How to Contribute to Open Source](https://opensource.guide/how-to-contribute/)" in the Open Source Guides.  \nYou can find personalized recommendations for projects and 

# ModelScope & FAISS

In [ ]:
'''
tracemalloc.start()
print(tracemalloc.get_traced_memory())

# Embeddings
from langchain_community.embeddings import ModelScopeEmbeddings
model_id = "damo/nlp_corom_sentence-embedding_english-base"
embeddings = ModelScopeEmbeddings(model_id=model_id)

# Store
docsearch = await FAISS.afrom_documents(docs, embeddings)

print(tracemalloc.get_traced_memory())
tracemalloc.stop()
'''

In [ ]:
'''
%%time
for i in range(10):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)
'''

# HuggingFaceEmbeddings & FAISS

In [3]:
tracemalloc.start()
print(tracemalloc.get_traced_memory())

# Embeddings
from langchain_community.embeddings import HuggingFaceEmbeddings
embeddings = HuggingFaceEmbeddings()

# Store
docsearch = await FAISS.afrom_documents(docs, embeddings)

print(tracemalloc.get_traced_memory())
tracemalloc.stop()

(928, 11633)
(611074318, 667287434)


In [8]:
%%time
for i in range(10):
    docsearch.similarity_search(query)

docsearch.similarity_search(query)

CPU times: user 2.29 s, sys: 69.1 ms, total: 2.36 s
Wall time: 619 ms


[Document(page_content="For more information on open source, specifically how to create and grow an open source project, we've created [Open Source Guides](https://opensource.guide/) that will help you foster a healthy open source community."),
 Document(page_content="For more information on open source, specifically how to create and grow an open source project, we've created [Open Source Guides](https://opensource.guide/) that will help you foster a healthy open source community by recommending best practices for creating and maintaining repositories for your open source project."),
 Document(page_content='Contributing to open source projects on {% data variables.product.prodname_dotcom %} can be a rewarding way to learn, teach, and build experience in just about any skill you can imagine. For more information, see "[How to Contribute to Open Source](https://opensource.guide/how-to-contribute/)" in the Open Source Guides.  \nYou can find personalized recommendations for projects and 

# Result 

| Embeddings Module     | Speed Create table (Cell Time) | Speed Search x10 (CPU) | Good Answer      | Memory Usage (Peak for 2K) |  Memory Usage (Peak for 18K) |
|-----------------------|--------------------------------|------------------------|:----------------:|----------------------------|------------------------------|
| BGE                   |                                |                        |                  | 185_254_579                | Not tested                   |
| FastEmbed             |                                |                        |                  | Max                        | Not tested                   |
| GPT4All               |32m 40.7s                       |2.96 s                  |🟢                | 31_415_910                 | 282854199                    |
| Hugging Face          |79m 58.6s                       |3.33 s                  |🟢                | 58_341_908                 | 667287104                    |
| ModelScope            |                                |                        |                  | 111_446_900                | Not tested                   |
| HuggingFaceEmbeddings |80m 2.1s                        |2.36 s                  |🟢                | 58_066_692                 | 667287434                    |
|                       |                                |                        |                  |                            |                              |

On the tested systems:
- HuggingFaceEmbeddings is the fastest for similarity search.
- GPT4All is the least resource-intensive in creating a store and will be more capable of avoiding exponential cloud computing costs.

